src/Microsoft.Diagnostics.Monitoring/DiagnosticsMonitor.cs

```diff
@@ -19,359 +19,119 @@ namespace Microsoft.Diagnostics.Monitoring
-                            {
-                                displayUnits = "count";
-                            }
-                            displayUnits += "/sec";
-                        }
-
-                        PostMetric(new Metric(traceEvent.TimeStamp, traceEvent.ProviderName, counterName, displayName, displayUnits, value, dimNames: DimNames, dimValues: _dimValues));
-                    }
-                }
-                catch (Exception ex)
-                {
-                    _logger.LogError(ex, "Error processing counter for {ProviderName}:{EventName}", traceEvent.ProviderName, traceEvent.EventName);
-                }
-            };
-        }
-
-        private void PostMetric(Metric metric)
-        {
-            foreach(IMetricsLogger metricLogger in _metricLoggers)
+            }
+            // On Unix platforms, we may actually get a PNSE since the pipe is gone with the process, and Runtime Client Library
+            // does not know how to distinguish a situation where there is no pipe to begin with, or where the process has exited
+            // before dotnet-counters and got rid of a pipe that once existed.
+            // Since we are catching this in StopMonitor() we know that the pipe once existed (otherwise the exception would've 
+            // been thrown in StartMonitor directly)
+            catch (PlatformNotSupportedException)
             {
-                try
-                {
-                    metricLogger.LogMetrics(metric);
-                }
-                catch (ObjectDisposedException)
-                {
-                }
-                catch (Exception e)
-                {
-                    _logger.LogError($"Error from {metricLogger.GetType()}: {e.Message}");
-                }
             }
         }
 
         public async ValueTask DisposeAsync()
         {
-            if (Interlocked.CompareExchange(ref _disposeState, 1, 0) == 1)
+            if (_disposed)
             {
                 return;
             }
-            
-            foreach(IMetricsLogger logger in _metricLoggers)
+
+            Task currentTask = null;
+            lock (_lock)
             {
-                if (logger is IAsyncDisposable asyncDisposable)
-                {
-                    await asyncDisposable.DisposeAsync();
-                }
-                else
+                if (_disposed)
                 {
-                    logger?.Dispose();
+                    return;
                 }
+                currentTask = _currentTask;
+                _currentTask = null;
+                _disposed = true;
             }
-        }
-
-        private class LogActivityItem
-        {
-            public Guid ActivityID { get; set; }
-
-            public LogObject ScopedObject { get; set; }
-
-            public LogActivityItem Parent { get; set; }
+             _disposeSource.Cancel();
+            if (currentTask != null)
+            {
+                await currentTask;
```

Comment:
I believe currentTask will be in the cancelled state if the cancellationToken that is passed into ProcessEvents is cancelled before the duration period expires. Thus, awaiting here will throw a TaskCanceledException, which I don't think you want to do in the DisposeAsync method.

Reply:
good catch.